stl/inc/algorithm

```diff
@@ -961,6 +941,322 @@ namespace ranges {
+        _NODISCARD static constexpr bool _Unsized_ranges(
+            _It1 _First1, _Se1 _End1, _It2 _First2, _Se2 _End2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            for (;; ++_First1, (void) ++_First2) { // Trim matching prefixes
+                if (_First1 == _End1) { // first range is a (possibly proper?) prefix of the second
+                    return _First2 == _End2;
+                } else if (_First2 == _End2) { // second range is a proper prefix of the first
+                    return false;
+                }
+
+                if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) { // mismatch
+                    break;
+                }
+            }
+            // If we get here, initial elements do not match.
+
+            // Determine final iterator values and validate lengths
+            auto _Last1 = _First1;
+            auto _Last2 = _First2;
+            for (;;) {
+                ++_Last1;
+                ++_Last2;
+                if (_Last1 == _End1) {
+                    if (_Last2 == _End2) {
+                        break; // equal lengths
+                    }
+
+                    return false; // different lengths; not a permutation
+                } else if (_Last2 == _End2) {
+                    return false; // Ditto different lengths
+                }
+            }
+            // if we get here, initial elements do not match and ranges have equal lengths.
+
+            if constexpr (bidirectional_iterator<_It1> && bidirectional_iterator<_It2>) {
+                for (;;) { // Trim matching suffixes
+                    if (--_Last1 == _First1) {
+                        return false; // initial elements did not match when trimming prefixes.
+                    }
+                    if (--_Last2 == _First2) {
+                        return false; // Ditto
+                    }
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_Last1), _STD invoke(_Proj2, *_Last2))) { // mismatch
+                        ++_Last1;
+                        ++_Last2;
+                        break;
+                    }
+                }
+                // If we get here, initial elements do not match, final elements do not match, and ranges have length
+                // at least 2.
+            }
+
+            return _Match_counts(
+                _STD move(_First1), _STD move(_Last1), _STD move(_First2), _STD move(_Last2), _Pred, _Proj1, _Proj2);
+        }
+
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Match_counts(const _It1 _First1, const _Se1 _Last1, const _It2 _First2,
+            const _Se2 _Last2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            for (auto _Mid1 = _First1; _Mid1 != _Last1; ++_Mid1) {
+                bool _Found = false;
+                auto _Meow  = _First1;

```

Comment:
As much as I like cat noises, we should choose a more descriptive name than `_Meow` for product code. `_Next` is often used for scanning.

Reply:
This is what I get for using a placeholder name other than `_Fixme`. `_Next` is generally used for scanning a forward range with a pair of adjacent iterators. I think `_Mid1` is really the perfect name for `_Meow`, and `_Current` is a better name for the iterator previously known as `_Mid1`.